pennylane/templates/subroutines/out_poly.py

```diff
@@ -0,0 +1,394 @@
+
+    def _flatten(self):
+        metadata1 = tuple((key, value) for key, value in self.hyperparameters.items())
+
+        return tuple(), metadata1
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+
+        hyperparams_dict = dict(metadata)
+        return cls(*data, **hyperparams_dict)
+
+    def map_wires(self, wire_map: dict):
+
+        new_input_registers = [
+            qml.wires.Wires([wire_map[wire] for wire in reg])
+            for reg in self.hyperparameters["input_registers"]
+        ]
+
+        new_output_wires = [wire_map[wire] for wire in self.hyperparameters["output_wires"]]
+
+        new_work_wires = (
+            [wire_map[wire] for wire in self.hyperparameters["work_wires"]]
+            if self.hyperparameters.get("work_wires")
+            else None
+        )
+
+        return OutPoly(
+            polynomial_function=self.hyperparameters["polynomial_function"],
+            input_registers=new_input_registers,
+            output_wires=new_output_wires,
+            mod=self.hyperparameters["mod"],
+            work_wires=new_work_wires,
+        )
+
+    @classmethod
+    def _primitive_bind_call(cls, *args, **kwargs):
+        return cls._primitive.bind(*args, **kwargs)
+
+    def decomposition(self):  # pylint: disable=arguments-differ
+        return self.compute_decomposition(**self.hyperparameters)
+
+    @staticmethod
+    def compute_decomposition(
+        polynomial_function, input_registers, output_wires, mod=None, work_wires=None
+    ):  # pylint: disable=unused-argument, arguments-differ
+        r"""Representation of the operator as a product of other operators (static method).
+
+        .. math:: O = O_1 O_2 \dots O_n.
+
+
+        .. seealso:: :meth:`~.OutPoly.decomposition`.
+
+        **Example:**
+
+        .. code-block:: python
+
+            print(
+            qml.OutPoly.compute_decomposition(
+                lambda x, y: x + y,
+                input_registers=[[0, 1],[2,3]],
+                output_wires=[4, 5],
+                mod=4,
+                )
+            )
+
+        .. code-block:: pycon
+
+        [QFT(wires=[4]), Controlled(PhaseAdder(wires=[4, None]), control_wires=[3]), Controlled(PhaseAdder(wires=[4, None]), control_wires=[1]), Adjoint(QFT(wires=[4]))]
+        """
+        registers_wires = [*input_registers, output_wires]
+
+        if not work_wires:
+            work_wires = [None, None]
+
+        list_ops = []
+
+        output_adder_mod = (
+            [work_wires[0]] + registers_wires[-1] if work_wires[0] else registers_wires[-1]
+        )
+
+        list_ops.append(qml.QFT(wires=output_adder_mod))
+
+        wires_vars = [len(w) for w in registers_wires[:-1]]
+
+        # Extract the coefficients and control wires from the binary polynomial
+        coeffs_list = _get_polynomial(polynomial_function, mod, *wires_vars)
+
+        all_wires_input = sum([*registers_wires[:-1]], start=[])
+
+        for item, coeff in coeffs_list.items():
```

Comment:
So the `coeffs` has to be integers (or the template works only with polynomials with integer coefficients)? Do we mention this somewhere in the docs?

Reply:
Oh good point! Now if the user defines functions with non-integer values, an error is raised. I added a test also